### PR TITLE
 Development of Interfaces, Mocks, and Unit Tests for Orchestrator and Related modules

### PR DESCRIPTION
**Pull request by juliancnn**
|Related issue|
|---|
|#20882|

### Overview

This PR introduces a comprehensive update to the Wazuh Engine, focusing on the development of interfaces, mocks, and unit tests for the Orchestrator, Worker, Tester, and Router components. These enhancements aim to improve the modularity, testability, and overall robustness of these key components within the Wazuh Engine.

### Key Features

- **Interface Development**: Implemented well-defined interfaces for the Orchestrator, Worker, Tester, and Router classes, separating them into distinct interfaces and implementations.
- **Mock Creation**: Developed detailed mocks for these interfaces, facilitating effective unit testing and simulation of various conditions and scenarios.
- **Orchestrator Unit Testing**: Created a comprehensive suite of unit tests for the Orchestrator, covering all its methods and interactions with Worker, Tester, and Router.
- **Integration with Existing Tests**: Ensured that these new unit tests seamlessly integrate with the existing test suite of the Wazuh Engine.

### Implementation Details

- Designed interfaces that encapsulate the functionalities of the Orchestrator, Worker, Tester, and Router, enhancing the separation of concerns and modularity.
- The mocks accurately mimic the behavior of the actual classes, including handling various inputs and states, to ensure effective unit testing.
- The unit tests for the Orchestrator are extensive, covering a wide range of scenarios, including edge cases and error handling.
- Followed established coding standards and best practices for the Wazuh Engine project throughout the development process.

### Testing and Validation

- Conducted rigorous unit testing to validate the functionality and reliability of the new interfaces and mocks.
- Ensured that the new unit tests do not disrupt the existing test suite and complement the current testing strategy.

### Impact

- This PR significantly enhances the testing infrastructure for the Orchestrator, Worker, Tester, and Router components in the Wazuh Engine, ensuring robustness and reliability through well-structured interfaces, mocks, and unit tests.

### CODE REVIEW

File: src/engine/source/router/test/src/unit/mockTester.hpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+                ingestTest,
+                (base::Event&&, const ::router::test::Options&),
+                (override));
+    MOCK_METHOD(base::RespOrError<std::unordered_set<std::string>>, getAssets, (const std::string&), (const, override));
+    MOCK_METHOD(bool, updateLastUsed, (const std::string&, uint64_t), (override));
+};
+
+} // namespace router
+
+#endif // ROUTER_TEST_INTERNAL_MOCKS_TESTER_HPP
```
Comment (JcabreraC) [inline src/engine/source/router/test/src/unit/mockTester.hpp:30]: The line break at the end of the file is missing.

File: src/engine/source/router/test/src/unit/mockWorker.hpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+public:
+    MOCK_METHOD(void, start, (), (override));
+    MOCK_METHOD(void, stop, (), (override));
+    MOCK_METHOD(const std::shared_ptr<IRouter>&, getRouter, (), (const, override));
+    MOCK_METHOD(const std::shared_ptr<ITester>&, getTester, (), (const, override));
+};
+
+} // namespace router
+
+#endif // ROUTER_TEST_INTERNAL_MOCKS_WORKER_HPP
```
Comment (JcabreraC) [inline src/engine/source/router/test/src/unit/mockWorker.hpp:25]: The line break at the end of the file is missing.

File: src/engine/source/router/test/src/unit/mockRouter.hpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+    MOCK_METHOD(base::OptError, enableEntry, (const std::string& name), (override));
+    MOCK_METHOD(base::OptError, changePriority, (const std::string& name, size_t priority), (override));
+    MOCK_METHOD(std::list<prod::Entry>, getEntries, (), (const, override));
+    MOCK_METHOD(base::RespOrError<prod::Entry>, getEntry, (const std::string& name), (const, override));
+    MOCK_METHOD(void, ingest, (base::Event && event), (override));
+};
+
+} // namespace router
+
+#endif // ROUTER_TEST_INTERNAL_MOCKS_ROUTER_HPP
```
Comment (JcabreraC) [inline src/engine/source/router/test/src/unit/mockRouter.hpp:25]: The line break at the end of the file is missing.

**Review by JcabreraC — CHANGES_REQUESTED** [review]

**Review by JavierBejMen — APPROVED** [review]
LGTM

**Review by soynof — CHANGES_REQUESTED** [review]

File: src/engine/source/router/CMakeLists.txt [hunk 4]
```diff
@@ -49,18 +49,33 @@ if(ENGINE_BUILD_TEST)
...
+    add_executable(router_ctest
+    ${COMPONENT_SRC_DIR}/router_test.cpp
+    ${COMPONENT_SRC_DIR}/tester_test.cpp
+    )
+    target_include_directories(router_ctest PRIVATE ${SRC_DIR})
+    target_link_libraries(router_ctest
+        gtest_main
+        base
+        router::router
+        builders # Todo delete after refactoring to builder
```
[the remark this answers is not in the document]
Comment (NahuFigueroa97) [reply src/engine/source/router/CMakeLists.txt:68]: yes

**Review by soynof — APPROVED** [review]
LGTM!
